R/preprocessing.R

```diff
@@ -645,6 +645,133 @@ Load10X_Spatial <- function (
+#' The CSV file should contain at least a "barcode" column that matches the cell barcodes in the Seurat object.
+#' Additional columns in the CSV file will be merged into the Seurat object's metadata.
+#'
+#' @importFrom utils read.csv
+#' @importFrom tibble rownames_to_column column_to_rownames
+#' @importFrom base file.path file.exists merge
+#'
+#' @examples
+#' \dontrun{
+#' # Specify the data directory containing the "cell_types" folder
+#' data.dir <- "/path/to/data"
+#'
+#' # Create a Seurat object (example)
+#' seurat_obj <- CreateSeuratObject(counts = some_counts_matrix)
+#'
+#' # Add cell type annotations to the Seurat object
+#' seurat_obj <- Add_10X_CellTypes(data.dir, seurat_obj)
+#' }
+
+Add_10X_CellTypes <- function(data.dir, object) {
+  cell_types_path <- file.path(data.dir, "cell_types", "cell_types.csv")
+  if (file.exists(cell_types_path)) {
+    cell.types <- read.csv(cell_types_path)
+    object@meta.data <- dplyr::left_join(tibble::rownames_to_column(object@meta.data, "barcode"),
+                                         cell.types, by = "barcode") %>%
+      tibble::column_to_rownames("barcode")
+    return(object)
+  } else {
+    return(object)
+  }
+}
+
+#' Load a 10x Genomics Single Cell Experiment into a \code{Seurat} object
+#'
+#' @inheritParams Read10X
+#' @inheritParams SeuratObject::CreateSeuratObject
+#' @param data.dir Directory containing the H5 file specified by \code{filename}
+#' and the image data in a subdirectory called \code{spatial}
+#' @param filename Name of H5 file containing the feature barcode matrix
+#' @param to.upper Converts all feature names to upper case. This can provide an
+#' approximate conversion of mouse to human gene names which can be useful in an
+#' explorative analysis. For cross-species comparisons, orthologous genes should
+#' be identified across species and used instead.
+#' @param ... Arguments passed to \code{\link{Read10X_h5}}
+#'
+#' @return A \code{Seurat} object
+#'
+#'
+#' @export
+#' @concept preprocessing
+#'
+#' @examples
+#' \dontrun{
+#' data_dir <- 'path/to/data/directory'
+#' list.files(data_dir) # Should show filtered_feature_bc_matrix.h5
+#' Load10X(data.dir = data_dir)
+#' }
+#'
+Load10X <- function(data.dir, filename = "filtered_feature_bc_matrix.h5",
+                    assay = "RNA", to.upper = FALSE, ...) {
+
+  if (length(data.dir) > 1) {
+    stop("`data.dir` expects a single directory path but received multiple values.")
+  }
+  if (!file.exists(data.dir)) {
+    stop(paste0("No such file or directory: '", data.dir, "'"))
+  }
+
+
+  filename <- list.files(data.dir, filename, full.names = FALSE, recursive = FALSE)
+  counts.path <- file.path(data.dir, filename)
+  if (!file.exists(counts.path)) {
+    stop(paste0("File not found: '", counts.path, "'"))
+  }
+
+  counts <- Read10X_h5(counts.path, ...)
+
+  if (to.upper) {
+    counts <- imap(counts, ~{
```

Comment:
Nit: because this formula operates on the counts matrices only and doesn't use the list's names, I think the `imap` call could be replaced by `base::sapply`.
Frankly, I'm not sure why this package imports `purrr::imap` anyway, as I can't find it used anywhere in the codebase, not even in `Load10X_Spatial`. But that's an issue for another time.

Reply:
personally, I like the simplicity and readability of tidy functions. I'll leave it up to the seruat developers to decide if they mind the extra import. I haven't heard any complaints since we wrote this function a few years ago.

Reply:
It's not really an issue; the choice one way or another doesn't affect me. (Personally, I agree; I always prefer the purrr functions in my own scripts.) I only wanted to mention it for the attention of the dev team.
Thanks for your receptiveness :smile: